datadog_checks_dev/datadog_checks/dev/tooling/test.py

```diff
@@ -0,0 +1,142 @@
+# (C) Datadog, Inc. 2018
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from .constants import TESTABLE_FILE_EXTENSIONS, get_root
+from .git import files_changed
+from .utils import get_testable_checks
+from ..subprocess import run_command
+from ..utils import chdir, path_join
+
+STYLE_ENVS = {
+    'flake8',
+}
+
+
+def get_tox_envs(checks, style=False, benchmark=False, every=False, changed_only=False, sort=False):
+    testable_checks = get_testable_checks()
+    changed_checks = get_changed_checks() if changed_only else None
+
+    if not checks:
+        checks = sorted(
+            testable_checks & (
+                get_changed_checks() if changed_checks is None else changed_checks
+            )
+        )
+
+    checks_seen = set()
+    for check in checks:
+        check, _, envs_selected = check.partition(':')
+
+        if (
+            check in checks_seen
+            or check not in testable_checks
+            or (changed_only and check not in changed_checks)
+        ):
+            continue
+        else:
+            checks_seen.add(check)
+
+        envs_selected = envs_selected.split(',') if envs_selected else []
+        envs_available = get_available_tox_envs(check, sort=sort)
+
+        if style:
+            envs_selected[:] = [
+                e for e in envs_available
+                if e in STYLE_ENVS
+            ]
+        elif benchmark:
+            envs_selected[:] = [
+                e for e in envs_available
+                if 'bench' in e
+            ]
+        else:
+            if every:
+                envs_selected[:] = envs_available
+            elif envs_selected:
+                available = set(envs_selected) & set(envs_available)
+                selected = []
+
+                # Retain order
+                for e in envs_selected:
```

Comment:
Can we do
```python
envs_selected[:] = [e for e in envs_selected if e in available]
``` 
?
Would be way simpler to read

Reply:
Can we really expect to have duplicates here ? Seems overkill to me to try and sanitize the input. I don't think it's worth the added complexity

Reply:
I think it is, for sane behavior. After all it's just 3 extra lines 🙂

Reply:
3 extra lines that make the code more intricate and hard to read. 
And it's not the behavior of tox either. You can run `tox -e unit,unit` and it will run the unit env twice.

Reply:
Tox may do that because we can't since we'll allow very soon the ability to keep envs up (docker, etc.) and multiple identical set up runs will fail or have other undesirable effects

Reply:
Tox may do that but we can't since we'll allow very soon the ability to keep envs up (docker, etc.) and multiple identical set up runs will fail or have other undesirable effects

Reply:
Fair enough !